mountpoint-s3/src/inode.rs

```diff
@@ -131,54 +131,62 @@ impl Superblock {
     /// The kernel may forget a number of references (`n`) in one forget message to our FUSE implementation.
     /// If the lookup count reaches zero, it is safe for the [Superblock] to delete the [Inode].
     pub fn forget(&self, ino: InodeNo, n: u64) {
-        let mut inodes = self.inner.inodes.write().unwrap();
-        match inodes.get(&ino) {
-            None => {
+        let inode = {
+            if let Some(inode) = self.inner.inodes.read().unwrap().get(&ino).cloned() {
+                inode
+            } else {
                 debug_assert!(
                     false,
                     "forget should not be called on inode already removed from superblock"
                 );
                 error!("forget called on inode {ino} already removed from the superblock");
+                return;
             }
-            Some(inode) => {
-                logging::record_name(inode.name());
-                let new_lookup_count = inode.dec_lookup_count(n);
-                if new_lookup_count == 0 {
-                    // Safe to remove, kernel no longer has a reference to it.
-                    trace!(ino, "removing inode from superblock");
-                    let inode = inodes.remove(&ino).expect("we just retrieved it");
+        };
+
+        logging::record_name(inode.name());
+        let new_lookup_count = inode.dec_lookup_count(n);
+        if new_lookup_count == 0 {
+            // Safe to remove, kernel no longer has a reference to it.
+            trace!(ino, "removing inode from superblock");
+            let Some(inode) = self.inner.inodes.write().unwrap().remove(&ino) else {
```

Comment:
Not blocking this PR to get merged. Just want to ask how scoping down the write lock is preventing the race condition?

Reply:
Do you mind to share the exact (or one of the) sequence of locks that could cause the deadlock, @monthonk?

Reply:
So, there are two things
1. The prevent deadlock, whenever we want to hold the lock for more than one resources we have to do it in the same sequence. But we did not follow this rule in previous code since 
    - `create` would lock `parent_state` first to add a child inode, then lock `inodes` to add the new inode to the hash map.
    - `forget` would lock `inodes` first because we want to remove a given inode from the hash map, then we lock `parent_state` to remove it from the parent inode.

a possible sequence of deadlock could be `create::lock::parent_state -> forget::lock::inodes -> create::lock::inodes -> forget::lock::parent_state`. We can fix this problem by holding one lock at a time for each operation.

2. Even if we can prevent the race condition from 1. we also don't want to hold the write lock on `inodes` longer than needed because this hash map is used across all file system operations. That's why we also scoped down the write lock in `forget` into multiple read/write locks. Basically, it's not necessary thing to do but a best practice.